File: JavaCivitas/src/civitas/CivitasJuego.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;
import java.util.ArrayList;

public class CivitasJuego {
    
    //solo se ponen los que salen no los que la flecha termina en la clase
    private int indiceJugadorActual;
    
    //relacion a jugador
    private ArrayList<Jugador> jugadores;
    //relacion Estados juegos
    private EstadosJuego estado;
    //relacion GestorEstados
    private GestorEstados gestorEstados;
    //relacion a Tablero
    private Tablero tablero = new Tablero(6);
    //private Tablero tablero;
    //relacion MAZOSORPRESAS
    private MazoSorpresas mazo;
    
    
    //metodos
    
    public void avanzaJugador(){
        Jugador jugadorActual = jugadores.get(indiceJugadorActual);
        int posicionActual = jugadorActual.getNumCasillaActual();
        int tirada = Dado.getInstance().tirar();
        int posicionNueva = tablero.nuevaPosicion(posicionActual, tirada);
        Casilla casilla;
        casilla = tablero.getCasilla(posicionNueva);
        this.contabilizarPasosPorSalida(jugadorActual);
        jugadorActual.moverACasilla(posicionNueva);
        //System.out.print("ANTES DE CASILLA");
        casilla.recibeJugador(indiceJugadorActual,jugadores);
        this.contabilizarPasosPorSalida(jugadorActual);
    }
    
    
    public boolean cancelarHipoteca(int ip){
        Jugador actual = jugadores.get(indiceJugadorActual);
        return actual.cancelarHipoteca(ip);
    }
    
    //constructor CIVITAS
    
    public CivitasJuego(ArrayList<String> nombres){

        ArrayList<Jugador> players = new ArrayList();
        for(int i=0; i<nombres.size(); i++){
            //añade los jugadores que se han insertado
            players.add(new Jugador(nombres.get(i)));
        }
        jugadores = players;
        gestorEstados = new GestorEstados();
        estado = EstadosJuego.INICIO_TURNO;
        indiceJugadorActual = Dado.getInstance().quienEmpieza(nombres.size()-1);
        mazo = new MazoSorpresas(true);
        inicializarMazoSorpresas(tablero);
        inicializarTablero(mazo);
    }
    
    public boolean comprar(){
        Jugador jugadorActual = jugadores.get(indiceJugadorActual);
        int numCasillaActual = jugadorActual.getNumCasillaActual();
        Casilla casilla = tablero.getCasilla(numCasillaActual);
        TituloPropiedad titulo = ((CasillaCalle)casilla).getTituloPropiedad();
        return jugadorActual.comprar(titulo);
    }
    
    public boolean construirCasa(int ip){
        Jugador jugadorActual = jugadores.get(indiceJugadorActual);
        return jugadorActual.construirCasa(ip);
    }
    
    public boolean construirHotel(int ip){
        Jugador jugadorActual = jugadores.get(indiceJugadorActual);
        return jugadorActual.construirHotel(ip);
    }
    
    private void contabilizarPasosPorSalida(Jugador jugadorActual){
        while(tablero.getPorSalida() > 0){
            jugadorActual.pasaPorSalida();
        }
    }
    
    public boolean finalDelJuego(){
        boolean finDelJuego=false;
        for(int i =0; i<jugadores.size() && !jugadores.get(i).enBancarrota(); i++){
            if(jugadores.get(i).enBancarrota()){
                finDelJuego=true;
            }
        }
        return finDelJuego;
    }
    
    public Casilla getCasillaActual(){
        int jugadorActual = jugadores.get(indiceJugadorActual).getNumCasillaActual();
        return tablero.getCasilla(jugadorActual);
    }
    
    public Jugador getJugadorActual(){
        return jugadores.get(indiceJugadorActual);
    }
    
    public boolean hipotecar(int ip){
        Jugador jugadorActual=jugadores.get(indiceJugadorActual);
        return jugadorActual.hipotecar(ip);
    }
    
    public String infoJugadorTexto(){
        return jugadores.get(indiceJugadorActual).toString();
    }
    
    private void inicializarMazoSorpresas(Tablero tablero){

        Sorpresa posicion = new SorpresaPosicion(4,tablero,25);
        mazo.alMazo(posicion);

        Sorpresa Especulador = new SorpresaEspeculador(1500,"[SORPRESA] Conversion ESPECULADOR");
        mazo.alMazo(Especulador);

        Sorpresa irACarcel = new SorpresaIrCarcel(tablero, "[SORPRESA] Ir Carcel");
        mazo.alMazo(irACarcel);

        Sorpresa irACasilla = new SorpresaIrCasilla(tablero,3,"[SORPRESA] Vete a la casilla 3");
        mazo.alMazo(irACasilla);

        Sorpresa irACasillaN2 = new SorpresaIrCasilla(tablero,5,"[SORPRESA] Vete a la casilla 5");
        mazo.alMazo(irACasillaN2);

        Sorpresa salirCarcel = new SorpresaSalirCarcel(mazo, "[SORPRESA] Salvoconducto obtenido");
        mazo.alMazo(salirCarcel);

        Sorpresa pagar = new SorpresaPagarCobrar(-50,"[SORPRESA] Tienes que pagar 50");
        mazo.alMazo(pagar);

        Sorpresa cobrar = new SorpresaPagarCobrar(50,"[SORPRESA] Tienes que cobrar 50");
        mazo.alMazo(cobrar);

        Sorpresa porCasaHotel = new SorpresaPorCasaHotel(20,"[SORPRESA] Pagas 100 por cada casa y hotel");
        mazo.alMazo(porCasaHotel);

        Sorpresa porJugador1 = new SorpresaPorJugador(30,"[SORPRESA] Cada jugador te paga 30");
        mazo.alMazo(porJugador1);

        Sorpresa porJugador2 = new SorpresaPorJugador(-30,"[SORPRESA] Se te quitan 30 para pagar a un jugador");
        mazo.alMazo(porJugador2);
    }
    
    public void inicializarTablero(MazoSorpresas mazo){

        tablero = new Tablero(6);

        // CASILLA 1
        TituloPropiedad titulo1 = new TituloPropiedad("Camino Purchil",100,(float)1.3,200,300,50); //Calle
        CasillaCalle casilla1=new CasillaCalle(titulo1);
        tablero.añadeCasilla(casilla1);

        // CASILLA 2
        TituloPropiedad titulo2 = new TituloPropiedad("Genil",85,(float)1.2,150,250,50);
        CasillaCalle casilla2=new CasillaCalle(titulo2);
        tablero.añadeCasilla(casilla2);

        // CASILLA 3
        TituloPropiedad titulo3 = new TituloPropiedad("Avenida Madrid",150,(float)1.4,250,350,75);
        CasillaCalle casilla3=new CasillaCalle(titulo3);
        tablero.añadeCasilla(casilla3);

        // CASILLA 4
        TituloPropiedad titulo4 = new TituloPropiedad("Camino Ronda",200,(float)1,100,440,55);
        CasillaCalle casilla4=new CasillaCalle(titulo4);
        tablero.añadeCasilla(casilla4);

        // CASILLA 5
        TituloPropiedad titulo5 = new TituloPropiedad("Reyes Catolicos",95,(float)1.9,120,490,20);
        CasillaCalle casilla5=new CasillaCalle(titulo5);
        tablero.añadeCasilla(casilla5);

        // CASILLA IMPUESTO
        CasillaImpuesto impuesto = new CasillaImpuesto(100,"Factura Endesa: 100"); //Impuesto
        tablero.añadeCasilla(impuesto);

        // CASILLA SORPRESA
        CasillaSorpresa sorpresa1 = new CasillaSorpresa(mazo,"Sorpresas");
        tablero.añadeCasilla(sorpresa1);

        // CASILLA 6
        TituloPropiedad titulo6 = new TituloPropiedad("Marbella",300,(float)1.5,1500,3500,980);
        CasillaCalle casilla6=new CasillaCalle(titulo6);
        tablero.añadeCasilla(casilla6);
        TituloPropiedad titulo7 = new TituloPropiedad("Caceres",400,(float)1.6,900,1500,150);
        CasillaCalle casilla7=new CasillaCalle(titulo7);
        tablero.añadeCasilla(casilla7);
        TituloPropiedad titulo8 = new TituloPropiedad("Gonzalo Gallas",500,(float)1.7,600,200,250);
        CasillaCalle casilla8=new CasillaCalle(titulo8);
        tablero.añadeCasilla(casilla8);
        TituloPropiedad titulo9 = new TituloPropiedad("Plaza Einstein",550,(float)1.8,3000,1200,500);
        CasillaCalle casilla9=new CasillaCalle(titulo9);
        tablero.añadeCasilla(casilla9);

        CasillaSorpresa sorpresa2 = new CasillaSorpresa(mazo, "Que Sera");
        tablero.añadeCasilla(sorpresa2);

        CasillaSorpresa sorpresa3 = new CasillaSorpresa(mazo, "Otra vez");
        tablero.añadeCasilla(sorpresa3);

        TituloPropiedad titulo10 = new TituloPropiedad("California",700,1.1f,1800,2500,1000);
        CasillaCalle casilla10=new CasillaCalle(titulo10);
        tablero.añadeCasilla(casilla10);
        TituloPropiedad titulo11 = new TituloPropiedad("Casablanca",500,1.3f,1200,1600,800);
        CasillaCalle casilla11=new CasillaCalle(titulo11);
        tablero.añadeCasilla(casilla11);
        TituloPropiedad titulo12 = new TituloPropiedad("Wellington",1200,1.4f,2000,2300,1000);
        CasillaCalle casilla12=new CasillaCalle(titulo12);
        tablero.añadeCasilla(casilla12);

        Casilla parking = new Casilla("Parking");
        tablero.añadeCasilla(parking);

        tablero.añadeJuez();
    }
    
    public void pasarTurno(){
        indiceJugadorActual = (indiceJugadorActual+1)% jugadores.size();
    }

    
    public ArrayList<Jugador> ranking(){
        ArrayList<Jugador> devolver = new ArrayList();
        ArrayList<Jugador> rank = new ArrayList(jugadores);
        
        Jugador jugadorTop;
        
        for(int i=0; i< jugadores.size(); i++){
            Jugador max = jugadores.get(1);
            for(int j=0; j< rank.size();j++){
                jugadorTop = jugadores.get(i);
                if(max.compareTo(jugadorTop) == 1){
                    max = jugadorTop;
                }
            }
            rank.remove(max);
            devolver.add(max);
        }
        return devolver;
    }
    
    public boolean salirCarcelPagando(){
        Jugador jugadorActual = jugadores.get(indiceJugadorActual);
        return jugadorActual.salirCarcelPagando();
    }
    
    public boolean salirCarcelTirando(){
        Jugador jugadorActual = jugadores.get(indiceJugadorActual);
        return jugadorActual.salirCarcelTirando();
    }
    
    public OperacionesJuego siguientePaso(){
       Jugador jugadorActual = jugadores.get(indiceJugadorActual);
       OperacionesJuego operacion = gestorEstados.operacionesPermitidas(jugadorActual, estado);
       if(operacion == OperacionesJuego.PASAR_TURNO){
           this.pasarTurno();
           this.siguientePasoCompletado(operacion);
       }
       if(operacion == OperacionesJuego.AVANZAR){
           this.avanzaJugador();
           this.siguientePasoCompletado(operacion);
       }
       contabilizarPasosPorSalida(jugadorActual);
       return operacion;
    }
    
    public void siguientePasoCompletado(OperacionesJuego operacion){
        Jugador jugador = jugadores.get(indiceJugadorActual);
        gestorEstados.siguienteEstado(jugador,estado,operacion);
    }
    
    public boolean vender(int ip ){
        Jugador actual = jugadores.get(indiceJugadorActual);
        return actual.vender(ip);
    }
    
    
    public void actualizarInfo(){
       
        System.out.println(this.infoJugadorTexto());
        
        for (int i = 0; i < jugadores.size() -1 && !jugadores.get(i).enBancarrota(); i++){
            if (jugadores.get(i+1).enBancarrota())
                ranking();
        }
    }
    
    
    
    
    
    
    
    
    
    
    
}

File: civitasruby/lib/civitas_juego.rb
# encoding:utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative ('tablero')
require_relative ('casilla')
require_relative ('gestor_estados')
require_relative ('titulo_propiedad')
require_relative ('jugador')
require_relative 'operaciones_juego'
require_relative 'casilla_calle'
require_relative 'casilla_juez'
require_relative 'casilla_impuesto'
require_relative 'casilla_sorpresa'
require_relative 'sorpresa'
require_relative 'sorpresa_ir_carcel'
require_relative 'sorpresa_ir_casilla'
require_relative 'sorpresa_pagar_cobrar'
require_relative 'sorpresa_por_casa_hotel'
require_relative 'sorpresa_por_jugador'
require_relative 'sorpresa_salir_carcel'
require_relative 'sorpresa_jugador_especulador'




module Civitas 
class CivitasJuego
  
  def initialize(nombres)
    @jugadores=Array.new
    for i in 0...nombres.size
      @jugadores<<Jugador.new(nombres[i])
    end
    @indiceJugadorActual=Dado.instance.quien_empieza(@jugadores.size)
    @gestorEstados=Gestor_estados.new
    @estado=@gestorEstados.estado_inicial
    @mazo=Mazo_sorpresas.new(false)
    inicializa_tablero(@mazo)
    inicializa_mazo_sorpresas(@tablero)
  end
  
  def inicializa_tablero(mazo)
    @tablero=Tablero.new(4)
    @mazo=mazo
    
    @tablero.anade_casilla(CasillaCalle.new("Albolotee",TituloPropiedad.new("Albolote",80,1.20,610,880,200)))
    @tablero.anade_casilla(CasillaSorpresa.new("Sorpresa1",@mazo))
    @tablero.anade_juez
    @tablero.anade_casilla(CasillaCalle.new("Ogijares",TituloPropiedad.new("Ogijares",70,1.20,590,840,185)))
    @tablero.anade_casilla(CasillaCalle.new("Armilla",TituloPropiedad.new("Armilla",60,1.17,565,800,170)))
    @tablero.anade_casilla(CasillaSorpresa.new("Sorpresa2",@mazo))
    
    @tablero.anade_casilla(CasillaCalle.new("Otura",TituloPropiedad.new("Otura",110,1.29,680,1150,240)))
    @tablero.anade_casilla(CasillaCalle.new("Chana",TituloPropiedad.new("Chana",100,1.26,650,950,230)))
    @tablero.anade_casilla(Casilla.new("Descanso"))
    @tablero.anade_casilla(CasillaCalle.new("Peligros",TituloPropiedad.new("Peligros",90,1.23,630,920,215)))
    @tablero.anade_casilla(CasillaSorpresa.new("Sorpresa3",@mazo))
    @tablero.anade_casilla(CasillaCalle.new("Churriana",TituloPropiedad.new("Churriana",170,1.44,830,1650,310)))
    @tablero.anade_casilla(CasillaCalle.new("Santa fe",TituloPropiedad.new("Santa fe",140,1.37,770,1500,280)))
    
    @tablero.anade_casilla(CasillaCalle.new("Baza",TituloPropiedad.new("Baza",130,1.34,740,1300,270)))
    @tablero.anade_casilla(CasillaSorpresa.new("Sorpresa4",@mazo))
    @tablero.anade_casilla(CasillaCalle.new("Cenes",TituloPropiedad.new("Cenes",210,1.53,950,2200,360)))
    @tablero.anade_casilla(CasillaCalle.new("Cullar vega",TituloPropiedad.new("Cullar vega",200,1.5,920,2000,350)))
    @tablero.anade_casilla(CasillaImpuesto.new("Multa",1000))
    @tablero.anade_casilla(CasillaCalle.new("Motril",TituloPropiedad.new("Motril",180,1.45,860,1850,320)))
   
    
  end
  
  
  def inicializa_mazo_sorpresas(tablero)
    @mazo.al_mazo(SorpresaJugadorEspeculador.new("200 de fianza", 200))
    @mazo.al_mazo(SorpresaPorJugador.new("Pagas 200 a cada jugador ", -200))
    @mazo.al_mazo(SorpresaSalirCarcel.new("Sorpresa salir carcel", @mazo))
    @mazo.al_mazo(SorpresaPagarCobrar.new("Cobra 500 euros",+500))
    @mazo.al_mazo(SorpresaPagarCobrar.new("Pagas 500 euros",-500))
    @mazo.al_mazo(SorpresaIrCasilla.new("Avanza hasta la casilla Salida", tablero, 0))
    @mazo.al_mazo(SorpresaIrCasilla.new("Vas a la carcel", tablero, 4))
    @mazo.al_mazo(SorpresaIrCasilla.new("Ve a la casilla de descanso", tablero,8)) 
    @mazo.al_mazo(SorpresaPorCasaHotel.new("Recibes 200 por cada casa y hotel", +200))
    @mazo.al_mazo(SorpresaPorCasaHotel.new("Pagas 200 por cada casa y hotel", -200))
    @mazo.al_mazo(SorpresaPorJugador.new("Recibes 200 de cada jugador ", +200))
    
    @mazo.al_mazo(SorpresaIrCarcel.new("Vas a la carcel",tablero))
  end
  
  def contabilizar_pasos_por_salida(jugadorActual)
    while(@tablero.get_porsalida>0)
      jugadorActual.pasa_por_salida
    end
  end
  
  def pasar_turno
    @indiceJugadorActual= @indiceJugadorActual+1
    if(@indiceJugadorActual==@jugadores.size)
      @indiceJugadorActual=0
    end
  end
  
   def siguiente_paso_completado(operacion)
     @estado=@gestorEstados.siguiente_estado(@jugadores[@indiceJugadorActual],
       @estado, operacion)
   end
   
   def vender(ip)
     return @jugadores[@indiceJugadorActual].vender(ip)
   end
   
   def salir_carcel_pagando
    return @jugadores[@indiceJugadorActual].salir_carcel_pagando
   end
   
   def salir_carcel_tirando
     return @jugadores[@indiceJugadorActual].salir_carcel_tirando
   end
   
   def final_del_juego
     ok=false
     for i in 0...@jugadores.size 
       if(@jugadores[i].en_bancarrota)
         ok=true
       end
     end
     return ok
   end
   
   def get_casilla_actual
     return @tablero.get_casilla(@jugadores[@indiceJugadorActual].numCasillaActual)
   end
   
   def get_jugador_actual
     return @jugadores[@indiceJugadorActual]
   end
   
   def ranking
    ordenados=[]
    for i in 0...@jugadores.size
    ordenados<<@jugadores[i]
    end
    ordenados.sort
    return ordenados
  end
   
   
   def avanza_jugador
     jugador_actual=@jugadores[@indiceJugadorActual]
     posicion_actual=jugador_actual.numCasillaActual
     tirada=Dado.instance.tirar
     posicion_nueva=@tablero.nueva_posicion(posicion_actual, tirada)
     casilla=@tablero.get_casilla(posicion_nueva)
     contabilizar_pasos_por_salida(jugador_actual)
     jugador_actual.mover_a_casilla(posicion_nueva)
     casilla.recibe_jugador(@indiceJugadorActual,@jugadores)
     contabilizar_pasos_por_salida(jugador_actual)
   end
   
   
   def siguiente_paso
     jugador_actual=@jugadores[@indiceJugadorActual]
     operacion=@gestorEstados.operaciones_permitidas(jugador_actual,@estado)
     if(operacion==Civitas::Operaciones_juego::PASAR_TURNO)
       pasar_turno
       siguiente_paso_completado(operacion)
     else if(operacion==Civitas::Operaciones_juego::AVANZAR)
        avanza_jugador
        siguiente_paso_completado(operacion)
     end
    end
    
    return operacion
   end
   
   
   
   def comprar
     jugador_actual=@jugadores[@indiceJugadorActual]
     num_casilla_actual=jugador_actual.numCasillaActual
     casilla=@tablero.get_casilla(num_casilla_actual)
     titulo=casilla.tituloPropiedad
     res=jugador_actual.comprar(titulo)
     return res
   end
   
   def info_jugador_texto
     return @jugadores[@indiceJugadorActual].to_string
   end
   
   
   def hipotecar(ip)
     return @jugadores[@indiceJugadorActual].hipotecar(ip)
   end
   
   def cancelar_hipoteca(ip)
     return @jugadores[@indiceJugadorActual].cancelar_hipoteca(ip)
   end
   
   def construir_casa(ip)
     return @jugadores[@indiceJugadorActual].construir_casa(ip)
   end
   
   def construir_hotel(ip)
     return  @jugadores[@indiceJugadorActual].construir_hotel(ip)
   end
   
   
   
   
  
  private :inicializa_tablero,:inicializa_mazo_sorpresas,:contabilizar_pasos_por_salida,:pasar_turno,
          :avanza_jugador
  
  end
end